Code/Tools/AssetProcessor/native/ui/BuilderInfoMetricsModel.cpp

```diff
@@ -0,0 +1,261 @@
+                if (item->GetJobCount() == 0)
+                {
+                    return QVariant();
+                }
+                return item->GetTotalDuration() / item->GetJobCount();
+            case aznumeric_cast<int>(Column::TotalDuration):
+                return item->GetTotalDuration();
+            // Other columns are sorted by Qt::DisplayRole immediately below
+            }
+        case Qt::DisplayRole:
+            switch (index.column())
+            {
+            case aznumeric_cast<int>(Column::Name):
+                return item->GetName();
+            case aznumeric_cast<int>(Column::JobCount):
+                return item->GetJobCount();
+            case aznumeric_cast<int>(Column::AverageDuration):
+                if (item->GetJobCount() == 0)
+                {
+                    return QVariant();
+                }
+                return DurationToQString(item->GetTotalDuration() / item->GetJobCount());
+            case aznumeric_cast<int>(Column::TotalDuration):
+                return DurationToQString(item->GetTotalDuration());
+            default:
+                break;
+            }
+        default:
+            break;
+        }
+
+        return QVariant();
+    }
+
+    QVariant BuilderInfoMetricsModel::headerData(int section, Qt::Orientation orientation, int role) const
+    {
+        if (orientation != Qt::Horizontal || role != Qt::DisplayRole)
+        {
+            return QVariant();
+        }
+        if (section < 0 || section >= aznumeric_cast<int>(Column::Max))
+        {
+            return QVariant();
+        }
+
+        switch (section)
+        {
+        case aznumeric_cast<int>(Column::Name):
+            return tr("Name");
+        case aznumeric_cast<int>(Column::JobCount):
+            return tr("Job Count");
+        case aznumeric_cast<int>(Column::AverageDuration):
+            return tr("Average Duration");
+        case aznumeric_cast<int>(Column::TotalDuration):
+            return tr("Total Duration");
+        default:
+            AZ_Warning("Asset Processor", false, "Unhandled BuilderInfoMetricsModel header %d", section);
+            break;
+        }
+        return QVariant();
+    }
+
+    QModelIndex BuilderInfoMetricsModel::parent(const QModelIndex& index) const
+    {
+        if (!index.isValid())
+        {
+            return QModelIndex();
+        }
+
+        auto currentItem = static_cast<BuilderDataItem*>(index.internalPointer());
+        auto parentItem = currentItem->GetParent();
+        auto rootItem = m_data->m_root.get();
+            
+        if (parentItem == rootItem || parentItem == nullptr)
+        {
+            return QModelIndex();
+        }
+
+        QModelIndex parentIndex = createIndex(parentItem->GetRow(), 0, parentItem);
+        Q_ASSERT(checkIndex(parentIndex));
+        return parentIndex;
+    }
+
+    void BuilderInfoMetricsModel::OnDurationChanged(BuilderDataItem* item)
+    {
+        while (item)
```

Comment:
we know there is 100% going to be Parent or Null here?

Reply:
I would say yes. 

`item` can only be the return value of `BuilderDataItem::UpdateOrInsertEntry`. This method can only return `nullptr` ([code](https://github.com/aws-lumberyard-dev/o3de/blob/c5h11oh/builderlist-treeview/Code/Tools/AssetProcessor/native/ui/BuilderDataItem.cpp#L85)), an existing pointer ([code](https://github.com/aws-lumberyard-dev/o3de/blob/c5h11oh/builderlist-treeview/Code/Tools/AssetProcessor/native/ui/BuilderDataItem.cpp#L94)), or a newly created pointer ([code](https://github.com/aws-lumberyard-dev/o3de/blob/c5h11oh/builderlist-treeview/Code/Tools/AssetProcessor/native/ui/BuilderDataItem.cpp#L103)). 

In the newly created pointer case: if the new operator failed, `nullptr` is returned; if the emplace_back fails, the method has no effect. Also, both of them throw exception on failure.

Therefore, there should be no possibility that the pointer is not null and invalid.

Reply:
For the `item` that is the parent of previous `item`, I am not that sure. 

`BuilderDataItem` saves parent and children as shared_ptr. However, when GetParent() and GetChildren() get called, raw pointers are returned. 

I will try to change that API to return shared_ptr.

Reply:
Resolved:
Eventually used shared_ptr to save children, and use weak_ptr to save parent pointer, since the child does not *own* the parent. Children can check if parent is still valid before accessing it, including here.